File: rust/tests/identity_test.rs
//! identity_test.rs
//!
//! Enum variant checks, struct construction, and fixture example checks
//! for CircleIdentity, RegisteredDevice, and IdentityTier.

use circle_ai::identity::{CircleIdentity, IdentityTier, RegisteredDevice};
use chrono::DateTime;
use serde::Deserialize;

// ─────────────────────────────────────────────────────────────────────────────
// Fixture deserialization helpers
// ─────────────────────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct IdentityFixture {
    identity_id: String,
    display_name: String,
    preferred_language: Option<String>,
    tier: String,
    device_ids: Vec<String>,
    created_at: String,
    last_seen_at: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct DeviceFixture {
    device_id: String,
    identity_id: String,
    platform: String,
    device_name: Option<String>,
    registered_at: String,
    last_active_at: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ExampleFixture {
    id: String,
    identity: IdentityFixture,
    devices: Vec<DeviceFixture>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Fixture {
    identity_tiers: Vec<String>,
    platforms: Vec<String>,
    examples: Vec<ExampleFixture>,
}

fn load_fixture() -> Fixture {
    let fixtures_dir = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .join("fixtures");
    let path = fixtures_dir.join("identity.json");
    let text = std::fs::read_to_string(&path)
        .unwrap_or_else(|e| panic!("Failed to read {:?}: {}", path, e));
    serde_json::from_str(&text).expect("Failed to parse identity.json")
}

fn tier_from_str(s: &str) -> IdentityTier {
    match s {
        "Anonymous" => IdentityTier::Anonymous,
        "Pseudonymous" => IdentityTier::Pseudonymous,
        "Verified" => IdentityTier::Verified,
        other => panic!("Unknown tier: {}", other),
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Enum tests
// ─────────────────────────────────────────────────────────────────────────────

#[test]
fn test_identity_tier_variants() {
    // All three variants exist and are distinct
    let tiers = [
        IdentityTier::Anonymous,
        IdentityTier::Pseudonymous,
        IdentityTier::Verified,
    ];
    assert_eq!(tiers[0], IdentityTier::Anonymous);
    assert_eq!(tiers[1], IdentityTier::Pseudonymous);
    assert_eq!(tiers[2], IdentityTier::Verified);
    assert_ne!(tiers[0], tiers[1]);
    assert_ne!(tiers[1], tiers[2]);
}

#[test]
fn test_identity_tier_ordering() {
    // Tier order: Anonymous < Pseudonymous < Verified
    assert!(IdentityTier::Anonymous < IdentityTier::Pseudonymous);
    assert!(IdentityTier::Pseudonymous < IdentityTier::Verified);
    assert!(IdentityTier::Anonymous < IdentityTier::Verified);
}

#[test]
fn test_fixture_tier_order() {
    let fixture = load_fixture();
    assert_eq!(fixture.identity_tiers, vec!["Anonymous", "Pseudonymous", "Verified"]);
}

// ─────────────────────────────────────────────────────────────────────────────
// Fixture example checks
// ─────────────────────────────────────────────────────────────────────────────

#[test]
fn test_fixture_has_three_examples() {
    let fixture = load_fixture();
    assert_eq!(fixture.examples.len(), 3);
}

#[test]
fn test_verified_multi_device_example() {
    let fixture = load_fixture();
    let example = fixture
        .examples
        .iter()
        .find(|e| e.id == "verified_multi_device")
        .expect("verified_multi_device example missing");

    let id_fix = &example.identity;
    assert_eq!(id_fix.identity_id, "a1b2c3d4-e5f6-7890-abcd-ef1234567890");
    assert_eq!(id_fix.display_name, "Sipho Dlamini");
    assert_eq!(id_fix.preferred_language.as_deref(), Some("zu"));
    assert_eq!(id_fix.tier, "Verified");
    assert_eq!(id_fix.device_ids.len(), 3);

    assert_eq!(example.devices.len(), 3);
    assert_eq!(example.devices[0].platform, "android");
    assert_eq!(example.devices[1].platform, "watch");
    assert_eq!(example.devices[2].platform, "windows");
}

#[test]
fn test_pseudonymous_single_device_example() {
    let fixture = load_fixture();
    let example = fixture
        .examples
        .iter()
        .find(|e| e.id == "pseudonymous_single_device")
        .expect("pseudonymous_single_device example missing");

    assert_eq!(example.identity.tier, "Pseudonymous");
    assert_eq!(example.devices.len(), 1);
    assert_eq!(example.devices[0].platform, "ios");
}

#[test]
fn test_anonymous_iot_example() {
    let fixture = load_fixture();
    let example = fixture
        .examples
        .iter()
        .find(|e| e.id == "anonymous_iot")
        .expect("anonymous_iot example missing");

    assert_eq!(example.identity.tier, "Anonymous");
    assert_eq!(example.identity.display_name, "Guest");
    assert!(example.identity.preferred_language.is_none());
    assert_eq!(example.devices.len(), 1);
    assert_eq!(example.devices[0].platform, "iot");
    assert!(example.devices[0].device_name.is_none());
}

// ─────────────────────────────────────────────────────────────────────────────
// Struct construction tests
// ─────────────────────────────────────────────────────────────────────────────

#[test]
fn test_circle_identity_construction() {
    let now = chrono::Utc::now();
    let identity = CircleIdentity::new(
        "test-id-001",
        "Test User",
        Some("en".to_string()),
        IdentityTier::Pseudonymous,
        vec!["device-001".to_string()],
        now,
        now,
    );

    assert_eq!(identity.identity_id, "test-id-001");
    assert_eq!(identity.display_name, "Test User");
    assert_eq!(identity.preferred_language.as_deref(), Some("en"));
    assert_eq!(identity.tier, IdentityTier::Pseudonymous);
    assert_eq!(identity.device_ids.len(), 1);
}

#[test]
fn test_registered_device_construction() {
    let now = chrono::Utc::now();
    let device = RegisteredDevice::new(
        "device-001",
        "identity-001",
        "android",
        Some("My Phone".to_string()),
        now,
        now,
    );

    assert_eq!(device.device_id, "device-001");
    assert_eq!(device.identity_id, "identity-001");
    assert_eq!(device.platform, "android");
    assert_eq!(device.device_name.as_deref(), Some("My Phone"));
}

#[test]
fn test_fixture_platforms() {
    let fixture = load_fixture();
    let expected_platforms = vec![
        "android", "ios", "windows", "macos", "linux", "web", "watch", "iot",
    ];
    assert_eq!(fixture.platforms, expected_platforms);
}

#[test]
fn test_all_fixture_examples_build_structs() {
    let fixture = load_fixture();
    for example in &fixture.examples {
        let id_fix = &example.identity;
        let tier = tier_from_str(&id_fix.tier);

        let created = DateTime::parse_from_rfc3339(&id_fix.created_at)
            .expect("invalid created_at")
            .with_timezone(&chrono::Utc);
        let last_seen = DateTime::parse_from_rfc3339(&id_fix.last_seen_at)
            .expect("invalid last_seen_at")
            .with_timezone(&chrono::Utc);

        let identity = CircleIdentity::new(
            &id_fix.identity_id,
            &id_fix.display_name,
            id_fix.preferred_language.clone(),
            tier,
            id_fix.device_ids.clone(),
            created,
            last_seen,
        );
        assert_eq!(identity.identity_id, id_fix.identity_id);

        for dev_fix in &example.devices {
            let reg_at = DateTime::parse_from_rfc3339(&dev_fix.registered_at)
                .expect("invalid registered_at")
                .with_timezone(&chrono::Utc);
            let last_active = DateTime::parse_from_rfc3339(&dev_fix.last_active_at)
                .expect("invalid last_active_at")
                .with_timezone(&chrono::Utc);

            let device = RegisteredDevice::new(
                &dev_fix.device_id,
                &dev_fix.identity_id,
                &dev_fix.platform,
                dev_fix.device_name.clone(),
                reg_at,
                last_active,
            );
            assert_eq!(device.device_id, dev_fix.device_id);
        }
    }
}

File: go/tests/identity_test.go
// identity_test.go
//
// Validates:
//   - IdentityTier constants exist and are ordered correctly.
//   - Three fixture examples from fixtures/identity.json can be decoded
//     and round-tripped through CircleIdentity / RegisteredDevice.

package circleai_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

// ---------------------------------------------------------------------------
// Fixture types
// ---------------------------------------------------------------------------

type identityFixture struct {
	IdentityTiers []string          `json:"identityTiers"`
	Examples      []identityExample `json:"examples"`
}

type identityExample struct {
	ID          string       `json:"id"`
	Description string       `json:"description"`
	Identity    identityJSON `json:"identity"`
	Devices     []deviceJSON `json:"devices"`
}

type identityJSON struct {
	IdentityID        string   `json:"identityId"`
	DisplayName       string   `json:"displayName"`
	PreferredLanguage *string  `json:"preferredLanguage"`
	Tier              string   `json:"tier"`
	DeviceIDs         []string `json:"deviceIds"`
	CreatedAt         string   `json:"createdAt"`
	LastSeenAt        string   `json:"lastSeenAt"`
}

type deviceJSON struct {
	DeviceID     string  `json:"deviceId"`
	IdentityID   string  `json:"identityId"`
	Platform     string  `json:"platform"`
	DeviceName   *string `json:"deviceName"`
	RegisteredAt string  `json:"registeredAt"`
	LastActiveAt string  `json:"lastActiveAt"`
}

func loadIdentityFixture(t *testing.T) identityFixture {
	t.Helper()
	path := filepath.Join(fixturesDir(t), "identity.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read identity.json: %v", err)
	}
	var fix identityFixture
	if err := json.Unmarshal(data, &fix); err != nil {
		t.Fatalf("failed to parse identity.json: %v", err)
	}
	return fix
}

func tierFromString(s string) circleai.IdentityTier {
	switch s {
	case "Anonymous":
		return circleai.IdentityTierAnonymous
	case "Pseudonymous":
		return circleai.IdentityTierPseudonymous
	case "Verified":
		return circleai.IdentityTierVerified
	default:
		panic("unknown tier: " + s)
	}
}

func mustParseTime(t *testing.T, s string) time.Time {
	t.Helper()
	ts, err := time.Parse(time.RFC3339, s)
	if err != nil {
		t.Fatalf("failed to parse time %q: %v", s, err)
	}
	return ts.UTC()
}

// ---------------------------------------------------------------------------
// Tier constant tests
// ---------------------------------------------------------------------------

func TestIdentityTierOrder(t *testing.T) {
	// Anonymous < Pseudonymous < Verified
	if circleai.IdentityTierAnonymous >= circleai.IdentityTierPseudonymous {
		t.Error("Anonymous should be less than Pseudonymous")
	}
	if circleai.IdentityTierPseudonymous >= circleai.IdentityTierVerified {
		t.Error("Pseudonymous should be less than Verified")
	}
}

func TestIdentityTierValues(t *testing.T) {
	fix := loadIdentityFixture(t)
	// Fixture declares 3 tier names; ensure all three map without panic.
	for _, name := range fix.IdentityTiers {
		_ = tierFromString(name)
	}
}

// ---------------------------------------------------------------------------
// Fixture example tests
// ---------------------------------------------------------------------------

func TestIdentityFixtureExamples(t *testing.T) {
	fix := loadIdentityFixture(t)

	if len(fix.Examples) < 3 {
		t.Fatalf("expected at least 3 examples, got %d", len(fix.Examples))
	}

	for _, ex := range fix.Examples {
		ex := ex
		t.Run(ex.ID, func(t *testing.T) {
			// Build CircleIdentity from fixture.
			identity := circleai.CircleIdentity{
				IdentityID:        ex.Identity.IdentityID,
				DisplayName:       ex.Identity.DisplayName,
				PreferredLanguage: ex.Identity.PreferredLanguage,
				Tier:              tierFromString(ex.Identity.Tier),
				DeviceIDs:         ex.Identity.DeviceIDs,
				CreatedAt:         mustParseTime(t, ex.Identity.CreatedAt),
				LastSeenAt:        mustParseTime(t, ex.Identity.LastSeenAt),
			}

			if identity.IdentityID == "" {
				t.Error("IdentityID must not be empty")
			}
			if identity.DisplayName == "" {
				t.Error("DisplayName must not be empty")
			}
			if len(identity.DeviceIDs) != len(ex.Devices) {
				t.Errorf("DeviceIDs count: got %d, want %d",
					len(identity.DeviceIDs), len(ex.Devices))
			}

			// Build RegisteredDevices from fixture.
			for _, dj := range ex.Devices {
				device := circleai.RegisteredDevice{
					DeviceID:     dj.DeviceID,
					IdentityID:   dj.IdentityID,
					Platform:     dj.Platform,
					DeviceName:   dj.DeviceName,
					RegisteredAt: mustParseTime(t, dj.RegisteredAt),
					LastActiveAt: mustParseTime(t, dj.LastActiveAt),
				}
				if device.DeviceID == "" {
					t.Error("DeviceID must not be empty")
				}
				if device.IdentityID != identity.IdentityID {
					t.Errorf("device.IdentityID %q != identity.IdentityID %q",
						device.IdentityID, identity.IdentityID)
				}
			}
		})
	}
}

// ---------------------------------------------------------------------------
// Specific example assertions
// ---------------------------------------------------------------------------

func TestIdentityFixture_Verified(t *testing.T) {
	fix := loadIdentityFixture(t)
	var ex *identityExample
	for i := range fix.Examples {
		if fix.Examples[i].ID == "verified_multi_device" {
			ex = &fix.Examples[i]
			break
		}
	}
	if ex == nil {
		t.Fatal("verified_multi_device example not found")
	}
	if ex.Identity.Tier != "Verified" {
		t.Errorf("tier: got %q, want Verified", ex.Identity.Tier)
	}
	if len(ex.Devices) != 3 {
		t.Errorf("device count: got %d, want 3", len(ex.Devices))
	}
	if ex.Identity.PreferredLanguage == nil || *ex.Identity.PreferredLanguage != "zu" {
		t.Errorf("preferredLanguage: got %v, want \"zu\"", ex.Identity.PreferredLanguage)
	}
}

func TestIdentityFixture_Anonymous(t *testing.T) {
	fix := loadIdentityFixture(t)
	var ex *identityExample
	for i := range fix.Examples {
		if fix.Examples[i].ID == "anonymous_iot" {
			ex = &fix.Examples[i]
			break
		}
	}
	if ex == nil {
		t.Fatal("anonymous_iot example not found")
	}
	if ex.Identity.Tier != "Anonymous" {
		t.Errorf("tier: got %q, want Anonymous", ex.Identity.Tier)
	}
	if ex.Identity.PreferredLanguage != nil {
		t.Errorf("preferredLanguage: got %v, want nil", ex.Identity.PreferredLanguage)
	}
	if len(ex.Devices) != 1 || ex.Devices[0].Platform != "iot" {
		t.Errorf("expected 1 iot device, got %v", ex.Devices)
	}
}
